controllers/subscription_controller.go

```diff
@@ -97,6 +113,87 @@ func (r *SubscriptionReconciler) Reconcile(ctx context.Context, _ ctrl.Request)
 	return ctrl.Result{}, nil
 }
 
+func getTargetNamespaces(olmPkgRecords []*OlmPkgRecord) []string {
+
+	var namespacesMap = make(map[string]bool)
+	var namespaces []string
+
+	for _, olmPkgRecord := range olmPkgRecords {
+		if !namespacesMap[olmPkgRecord.Namespace] {
+			namespaces = append(namespaces, olmPkgRecord.Namespace)
+		}
+		namespacesMap[olmPkgRecord.Namespace] = true
+	}
+
+	return namespaces
+}
+
+func ensureNamespace(ctx context.Context, cli client.Client, logger logr.Logger, olmPkgRecords []*OlmPkgRecord) error {
+
+	namespacesMap := make(map[string]bool)
+
+	for _, olmPkgRecord := range olmPkgRecords {
+
+		if olmPkgRecord.Namespace == "" || olmPkgRecord.Namespace == OperatorNamespace {
+			logger.Info("skipping namespace creation", "record", olmPkgRecord)
+			continue
+		}
+
+		if !namespacesMap[olmPkgRecord.Namespace] {
+			ns := &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: olmPkgRecord.Namespace,
+				},
+			}
+			if err := cli.Create(ctx, ns); err != nil && !errors.IsAlreadyExists(err) {
+				logger.Error(err, "failed to create namespace", "namespace", olmPkgRecord.Namespace)
+				return err
+			} else {
+				logger.Info("created namespace", "namespace", olmPkgRecord.Namespace)
+			}
+
+			namespacesMap[olmPkgRecord.Namespace] = true
+		}
+	}
+
+	return nil
+}
+
+func ensureOperatorGroup(ctx context.Context, cli client.Client, logger logr.Logger, olmPkgRecords []*OlmPkgRecord) error {
+
+	namespacesMap := make(map[string]bool)
+
+	for _, olmPkgRecord := range olmPkgRecords {
+
+		if olmPkgRecord.Namespace == "" || olmPkgRecord.Namespace == OperatorNamespace {
+			logger.Info("skipping operatorGroup creation", "record", olmPkgRecord)
+			continue
+		}
+
+		if !namespacesMap[olmPkgRecord.Namespace] {
+			opGroup := &opv1.OperatorGroup{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      olmPkgRecord.Namespace + "-operator-group",
+					Namespace: olmPkgRecord.Namespace,
+				},
+				Spec: opv1.OperatorGroupSpec{
+					TargetNamespaces: []string{olmPkgRecord.Namespace},
```

Comment:
is this a requirement or an assumption for us to start w/ supporting single ns operators?

Reply:
assumption. Also we can not know if they need more then one namespace in advance. In that case anyone can go and edit it manually.